frontend/app/routes/$lang+/_protected+/personal-information+/home-address+/edit.tsx

```diff
@@ -82,40 +85,85 @@ export async function action({ context: { session }, params, request }: ActionFu
+          ctx.addIssue({ code: z.ZodIssueCode.custom, message, path: ['postalCode'] });
+        }
+      }
+    })
+    .transform((val) => {
+      return {
+        ...val,
+        postalCode: val.countryId && val.postalCode ? formatPostalCode(val.countryId, val.postalCode) : val.postalCode,
+      };
+    });
+
+  const formData = await request.formData();
 
-  const formData = Object.fromEntries(await request.formData());
   const expectedCsrfToken = String(session.get('csrfToken'));
-  const submittedCsrfToken = String(formData['_csrf']);
+  const submittedCsrfToken = String(formData.get('_csrf'));
 
   if (expectedCsrfToken !== submittedCsrfToken) {
     log.warn('Invalid CSRF token detected; expected: [%s], submitted: [%s]', expectedCsrfToken, submittedCsrfToken);
     throw new Response('Invalid CSRF token', { status: 400 });
   }
 
-  const parsedDataResult = await formDataSchema.safeParseAsync(formData);
+  const data = {
+    streetName: String(formData.get('streetName') ?? ''),
+    secondAddressLine: formData.get('secondAddressLine') ? String(formData.get('secondAddressLine')) : undefined,
+    countryId: String(formData.get('countryId') ?? ''),
+    provinceTerritoryStateId: formData.get('provinceTerritoryStateId') ? String(formData.get('provinceTerritoryStateId')) : undefined,
+    cityName: String(formData.get('cityName') ?? ''),
+    postalCode: formData.get('postalCode') ? String(formData.get('postalCode')) : undefined,
+    homeAndMailingAddressTheSame: formData.get('homeAndMailingAddressTheSame') === 'copy',
+  };
+  const parsedDataResult = formDataSchema.safeParse(data);
 
   if (!parsedDataResult.success) {
-    instrumentationService.countHttpStatus('home-address.edit', 400);
-    return json({
-      errors: parsedDataResult.error.flatten(),
-      formData: formData as Partial<z.infer<typeof formDataSchema>>,
-    });
+    return json({ errors: parsedDataResult.error.format() });
   }
+
   const newHomeAddress = parsedDataResult.data;
   session.set('newHomeAddress', newHomeAddress);
 
+  const { streetName, cityName, countryId, postalCode, provinceTerritoryStateId } = parsedDataResult.data;
+  const correctedAddress = await wsAddressService.correctAddress({ address: streetName, city: cityName, province: provinceTerritoryStateId, postalCode: postalCode, country: countryId });
+
   instrumentationService.countHttpStatus('home-address.edit', 302);
+  if (correctedAddress.status === 'Corrected') {
+    const { address, city, province, postalCode, country } = correctedAddress;
+    session.set('suggestedAddress', { address, city, province, postalCode, country });
+    return redirect(getPathById('$lang+/_protected+/personal-information+/home-address+/suggested', params));
+  }
+
+  if (correctedAddress.status === 'NotCorrect') {
+    return redirect(getPathById('$lang+/_protected+/personal-information+/home-address+/address-accuracy', params));
+  }
```

Comment:
We should remove these WSAddress references. They were previously removed in https://github.com/DTS-STN/canadian-dental-care-plan/pull/950/

Reply:
merged the changes